ISLANDORA-1402: URL Raw encode URL-breaking characters in form names

d-r-p opened the pull request:
**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-1402

* https://jira.duraspace.org/browse/ISLANDORA-1192 (original of the above duplicate; closed on 2018-05-25 for de-duplication)

# Note [introduced in edit]

The title of this PR
> ISLANDORA-1402: Forbid URL-breaking characters in form names as a temporary solution

should now read
> ISLANDORA-1402: Encode form names in URLs

# What does this Pull Request do?

[edit]

~Forbid '/', '#' and '?' in form names for newly created forms.~

Encode form names in URLs so that problematic characters (chiefly '`/`', '`#`' and, perhaps, even '`?`') do not break form accessibility (_Copy_/_Edit_/_View_/_Export_/_Delete_/_Associate_ links on `/admin/islandora/xmlform/forms`).

# What's new?

[edit]

~Add logic inside name validator function to throw form error when the submitted form name contains one of the aforementioned characters.~

`rawurlencode()` form names in displayed links and `rawurldecode()` form names in page callbacks.

# How should this be tested?

[edit]

On a test VM:

* ~Go to `http://localhost:8000/admin/islandora/xmlform/forms/create`~
* ~Introduce a form name that contains '`/`', '`#`' and/or '`?`'~
* ~Submit and (probably) find your form inaccessible, since the form name will be part of the URL for all tasks (copy, edit, view, export, delete, associate)~
* ~Try the same when copying a form (`http://localhost:18000/admin/islandora/xmlform/forms/YOUR_FORM_NAME/copy`)~
* ~Pull this PR and repeat~
* ~Confirm that you cannot create or copy a form giving it a name containing any of the aforementioned characters~

* Create, copy, view, export, delete and associate forms with any names and confirm that
    * form names are always encoded in the URLs
    * form names are displayed un-encoded
    * no hiccup occurs

# Additional Notes:

[edit]

~This is just a _temporary_ solution to [ISLANDORA-1402](https://jira.duraspace.org/browse/ISLANDORA-1402). Ultimately, we wish to modify the way form names are stored (probably giving them machine names that will serve as URL parts, while retaining full names for display).~

~Also note that we make no attempt here to "revive" forms that were created with problematic names. These can be considered lost for the time being.~

This PR modifies a few functions. In particular, the page callbacks of `xml_form_builder` now expect `rawurlencode()`d form names as parameter. This has the potential to break outside code that recycles those functions.

* Does this change require documentation to be updated? **probably not**
* Does this change add any new dependencies? **no**
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)? **no**
* Could this change impact execution of existing code? ~**no**~ **possibly**

# Interested parties
@dmoses @bondjimbond @willtp87 @adam-vessey @DiegoPino @Islandora/7-x-1-x-committers

d-r-p committed:
ISLANDORA-1402: Forbid URL-breaking characters in form names as a temporary solution.

Forbid '/', '#' and '?' in form names for newly created forms.

d-r-p committed:
ISLANDORA-1402 (continued)

Also forbid '/', '#' and '?' in form names for copied forms.

d-r-p commented:
Did not realise that _Copy_ uses its own validator... Update[d] ~coming up~...

jonathangreen commented:
Perhaps we should be using drupal machine names: 
https://api.drupal.org/api/drupal/includes%21form.inc/function/form_validate_machine_name/7.x

d-r-p commented:
@jonathangreen Yes, that is the ultimate goal but needs re-coding of the form repository (or shall we discard display names altogether?). Hence, just for the time being (release), a _temporary_ solution that will prevent the bug from biting users...

rosiel commented:
Could probably do "light sanitization" to remove problematic characters so they don't fail in the URLs? Something that will take care of 'dirty' characters while not changing (too much) existing forms that may have been bookmarked...

d-r-p commented:
@rosiel I am not sure how people would have bookmarked "dirty" forms (at least with '`/`' and '`#`' the forms should have been inaccessible, as Drupal parses the URLs differently; about '`?`' I am unsure - included it upon @adam-vessey's suggestion). Also, this PR only prevents creation of new form names with problematic characters - the old ones are unaffected. I do not know if this answers your question, or if I misunderstood something...

d-r-p committed:
ISLANDORA-1402: Encode form names in URLs instead of banning characters

- revert the validation against '/', '#' and '?' in form names
- rawurlencode form names in URLs (and rawurldecode them for usage)

Note: This might break logic for code that is recycling the page
callbacks of xml_form_builder.

While at it, add breadcrumbs to the associations form, too.

d-r-p commented:
Ok folks, since I got the drift that banning problematic characters until we re-code the form repository for good does not get a lot of support, I tried to provide an alternative (albeit a little more intrusive code-wise) solution by `rawurlencode()`ing form names _only in URLs_, and `rawurldecode()`ing them for use.

From my tests, users should not be affected too much by this solution (only _bookmarked_ form links might need updating), and problematic forms should even get resurrected this way.

I count on you experts to assess whether or not this is problematic for the code base (some functions now expect a `rawurlencode()`d parameter).

d-r-p commented:
Pinging @Islandora/7-x-1-x-committers: anyone who would like to give their two cents?

DiegoPino commented:
@d-r-p ok this works. I just need to give it a run since this changes some existing functions expectations (encoded v/s not) and there could be some outside invocations to them in the wild? (just a guess). Give me 24 hours. (which will then make this 27 days... sorry man)

DiegoPino commented:
@d-r-p quick thought... so... in case i do find some issues with change of function signature expectations (well... not signature.. more the value format right?) would simply doing the encoding/decoding inside the said functions a solution? instead of expecting the caller to do the job? That way we can keep all this around. Would of course be an overkill of decodes and encodes all over the place...

d-r-p commented:
Thank you, @DiegoPino, for taking a look!

Not sure about your proposition, though, since we are talking callback functions that get parts of the URI passed as parameter. As an alternative, I could introduce intermediary callbacks that do the translation and then call the original one. Would that work for you?

DiegoPino commented:
Hi, i mean your code works already for me! Please don’t changed anything
yet. I’m just preparing for the rare case where we have to maintain
backwards compat for some external caller, but to be honest, seems more
than improbable after looking around today

El El mié, 18 de jul. de 2018 a las 16:35, d-r-p <notifications@github.com>
escribió:

> Thank you, @DiegoPino <https://github.com/DiegoPino>, for taking a look!
>
> Not sure about your proposition, though, since we are talking callback
> functions that get parts of the URI passed as parameter. As an alternative,
> I could introduce intermediary callbacks that do the translation and then
> call the original one. Would that work for you?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Islandora/islandora_xml_forms/pull/257#issuecomment-406065497>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGn855MN8oLOwedm9hSg8qZ32mfylyWFks5uH5wMgaJpZM4UnCdH>
> .
>
-- 
Diego Pino Navarro
Digital Repositories Developer
Metropolitan New York Library Council (METRO)

DiegoPino reviewed (approved):
Cool. Vanilla Islandora should not complain about any of these. Probably some Bookmarked XML forms could need updates, but hey, how many bookmarks people keep? Great work. Will merge in 24

DiegoPino commented:
@d-r-p merging. Will do a squash here and will also change your Pull request title. No longer "forbid" but now it is "URL encode". is that OK?

d-r-p commented:
@DiegoPino Thank you very much! I would suggest removing "_as a temporary solution_" in the title, or do you think the original issue is not fully solved yet? Maybe "_ISLANDORA-1402: Raw encode form names in URLs_" (at your discretion with or without a trailing "_as a temporary solution_") would, actually, be more descriptive, since _all_ characters are raw encoded.
